drivers/nutdrv_qx.c

```diff
@@ -1952,10 +1951,10 @@ static int	armac_command(const char *cmd, char *buf, size_t buflen)
 			break;
 		}
 
-		if (bytes_available > ARMAC_READ_SIZE - 1) {
+		// if (bytes_available > ARMAC_READ_SIZE - 1) {
```

Comment:
If I read it right with current implementation `bytes_available` cannot be greater than 15. But we get more bytes for Q1 response and this always truncates final response

Reply:
Yeah, we've never figured out the meaning of the high nibble, but it can encode longer messages. 

Question is - how do one distinguish between those two protocols. The USB ID is probably the same. Trying message and falling back to interrupt might be slower and might confuse the UPS firmware. Maybe one can do it ONCE, and store the result somewhere (global static? uhm. Maybe driver has some state that can be used).

Maybe having two drivers armac and armac2 is viable, they could share some/most of the code. But the user will have to distinguish and select one.

As for the details - it certainly can be done. Would be best to have a PCAP. We have one here too:
https://github.com/networkupstools/nut/issues/2099
I had some local ugly `mock` that allowed me to run the code against stored messages. It's getting somewhat hard to have a nice regression testing without a proper test suite and USB mocking.

Reply:
`tmpbuf[0] & 0x3F` works great with current ups - thanks! c395c7568572163eea62bfd67eeaec737ceb34af

> BTW, maybe 64+1 for the string's final null byte could be a safer choice, remaining pre-zeroed from that memset() above, just in case :)

Not sure if I understand it correctly but probably 877913a4a4673d02f5d46b29dfd0b02d8d355410 might be the fix?

Reply:
> Question is - how do one distinguish between those two protocols

If `lsusb -v` actually shows correct endpoint address `bEndpointAddress` and transfer type `Transfer Type` then probably we can use use them in code. And then we might survive with single driver. libusb should expose them too https://stackoverflow.com/a/16438105

If you agree with this approach I could try to implement it. Maybe you could advice in which function would be best to retrieve device descriptors (upsdrv_initups, upsdrv_initinfo, something else)?

Reply:
Thinking of it, this would either way be a toggle to choose one of two otherwise similar behaviors. For the first approximation (and even later, for when smart ideas misfire) a manual toggle as a driver config option is a reasonable choice.

Just that later if it is default, we can try to guess the best setting at run-time.

Reply:
If we use libusb functions direct there then maybe we need to read struct: https://libusb.sourceforge.io/api-1.0/structlibusb__endpoint__descriptor.html#a111d087a09cbeded8e15eda9127e23d2

bmAttributes: Bits 0:1 determine the transfer type and correspond to [libusb_endpoint_transfer_type](https://libusb.sourceforge.io/api-1.0/group__libusb__desc.html#ga8c93c0ec3fccf1c44e8ec7639619a53e)

So we either have it somewhere, or need to read it. Probably starting with libusb_get_device_descriptor() or maybe libusb_get_active_config_descriptor?

Reply:
I've checked; there's only subdriver 'command' function. So to keep it "contained" you could store it there but there's no proper state to save it.

Drivers work as separate processes AFAIK. So there should be no situation in which one creates subdriver for one UPS and then reuses it in another. So static field in armac_command... could be fine maybe?

Otherwise you'd need to change qx driver API somehow. Create subdriver-specific state pass it around... I don't think that's a good idea.

Reply:
I have added `libusb_get_active_config_descriptor` and cache it in static variable. I enable new interrupt logic only when config fully matches new Armac ups. This way we hopefully won't break older ups models that works with `usb_control_msg`. 

I have not added the toggle as driver config option. Cannot figure what would be correct name for it. And now I'm not sure should it be just boolean or also it should describe in and out endpoint addresses and max packet size?